refactor(tests): Use pytest collection to load JSON fixtures

marioevz opened the pull request:
## 🗒️ Description
This PR refactors the blockchain and state test infrastructure to leverage pytest's native collection mechanism via pytest_collect_file, eliminating redundant JSON file reads and improving test execution efficiency.

#### Key Improvements

1. Native pytest Collection
  - Implements pytest_collect_file hook to collect tests directly from JSON files during pytest's discovery phase
  - Each JSON file is now read exactly once during collection, rather than being read multiple times during parameterization and execution
  - Test fixtures are created as pytest Item objects (e.g., BlockchainTestFixture, StateTestFixture) that encapsulate all test data

2. Eliminated Redundant File I/O
  - Before: JSON files were read during test parameterization (fetch_blockchain_tests) and again during test execution (run_blockchain_st_test)
  - After: JSON files are read once in FixturesFile.collect(), and test data is stored in fixture objects for later execution
  - Removes intermediate dictionaries passing file paths that triggered repeated file reads

3. Cleaner Architecture
  - Introduces Fixture base class for shared fixture behavior
  - Test execution logic moved into runtest() methods of fixture classes
  - Test metadata (markers, fork info) configured during collection rather than parameterization
  - Eliminates the need for custom idfn functions - pytest handles naming automatically

#### Performance Impact

This refactoring significantly reduces I/O overhead for large test suites where the same JSON files contain multiple test cases across different forks.

#### Open Issues

Some failing tests still that need to be investigated, for now I'd like to start running this in CI and see how it improves execution speed.

## 🔗 Related Issues or PRs
N/A.

## ✅ Checklist
- [x] All: Ran fast `tox` checks to avoid unnecessary CI fails, see also [Code Standards](https://eest.ethereum.org/main/getting_started/code_standards/) and [Enabling Pre-commit Checks](https://eest.ethereum.org/main/dev/precommit/):
    ```console
    uvx --with=tox-uv tox -e static
    ```
- [x] All: PR title adheres to the [repo standard](https://eest.ethereum.org/main/getting_started/contributing/?h=contri#commit-messages-issue-and-pr-titles) - it will be used as the squash commit message and should start `type(scope):`.
- [ ] All: Considered adding an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [x] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [ ] All: Set appropriate labels for the changes (only maintainers can apply labels).

#### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->]()

marioevz committed:
refactor(tests): Refactor json_infra using `pytest_collect_file`

marioevz committed:
fix(tests): json collecting

marioevz committed:
fix(tests): blockchain test execution

marioevz committed:
fix(tests): blockchain test execution

marioevz committed:
refactor(tests): Refactor types in json_infra

marioevz committed:
fix(tests): json_infra, imports, parse `exceptions` in some tests

marioevz committed:
refactor(tests): move some definitions

marioevz committed:
fix(tox.ini): Remove `--ignore-glob`

marioevz committed:
fix(tests): workaround for FileNotFoundError

SamWilsn commented on tests/json_infra/conftest.py:
```diff
@@ -260,6 +267,17 @@ def pytest_sessionstart(session: Session) -> None:
                     fixture_path,
                 )
 
+            # Remove any python files in the downloaded files to avoid
+            # importing them.
+            for python_file in glob(
+                os.path.join(fixture_path, "**/*.py"), recursive=True
+            ):
+                try:
+                    os.unlink(python_file)
+                except FileNotFoundError:
+                    # Not breaking error, another process deleted it first
+                    pass
+
```
This feels... strange? I can't quite put my finger on why.

Like, why do the fixtures contain python files at all? Is there another way we could accomplish the same thing (like excluding a directory)?

I dunno, this just triggers my spidey sense :rofl:

marioevz replied:
This is the culprit: https://github.com/ethereum/legacytests/tree/1f581b8ccdc4c63acf5f2c5c1b155c690c32a8eb/src/LegacyTests/Cancun/GeneralStateTestsFiller/Pyspecs

Checking out `ethereum/tests` at this commit, when submodules are included, results in these python files being checked out too, and when collecting `./tests/json_infra/fixtures` for JSON files, pytest tries to collect these files too.

SamWilsn replied:
Don't we exclude that directory on the command line?

marioevz replied:
I removed that because with this approach the files are collected directly by pytest, as opposed to doing a glob in the test itself.

SamWilsn commented on tests/json_infra/helpers/__init__.py:
```diff
@@ -1 +1,10 @@
+from .load_blockchain_tests import BlockchainTestFixture
+from .load_state_tests import StateTestFixture
+
+ALL_FIXTURE_TYPES.append(BlockchainTestFixture)
+ALL_FIXTURE_TYPES.append(StateTestFixture)
```
Do these get executed when importing only, for example, `.load_state_tests`? From my limited knowledge of Python's import machinery, I would guess yes, but I'm just checking.

marioevz replied:
Yes that's correct, it gets executed only when importing from `.helpers`. If we were to, for example, import directly from `.helpers.fixtures`, this logic would not be executed and `ALL_FIXTURE_TYPES` would be empty, so it is indeed a bit brittle if being honest.

SamWilsn replied:
Oh really? I thought parent modules were implicitly imported. I'm glad I checked!

SamWilsn commented on tests/json_infra/helpers/exceptional_test_patterns.py:
```diff
@@ -20,6 +21,7 @@ class TestPatterns:
     big_memory: Tuple[Pattern[str], ...]
 
 
+@lru_cache
```
How often is this called to require an `lru_cache`? O.o

Depending on when the cache is populated (in worker vs. in master), using `lru_cache` can explode memory: each worker has its own cache.

marioevz replied:
I removed it thinking it might reduce the memory footprint and it did by half a GB, but it still consumes around 30GB+ because all fixtures are in memory when running.

SamWilsn replied:
<img width="512" height="512" alt="image" src="https://github.com/user-attachments/assets/4937760e-9a55-43ae-8d39-fcdcdf25ed9b" />

marioevz committed:
fix(tests): revamp cache

fix(tests): Don't cache fixtures

Try to implement cache

Fix caching

feat(tests): Manage cache during execution

SamWilsn commented:
I was thinking briefly about this. I also know next to nothing about pytest, so this might not make any sense at all, but...

What if we use an LRU cache for the JSON files (one per worker), and [loadgroup](https://pytest-xdist.readthedocs.io/en/latest/distribution.html#distribution-modes) all the tests that come from the same file?

So you'd read once during collection, find all the tests and group them by file, then while running the tests you minimize the number of times you need to re-read the same file.

marioevz committed:
fix(tox): Use `--dist=loadfile`

marioevz committed:
fix(tests): json files cache

gurukamath commented on tests/json_infra/helpers/load_blockchain_tests.py:
```diff
@@ -133,96 +58,150 @@ def add_block_to_chain(
+        super().__init__(*args, **kwargs)
+        self.fork_name = self.test_dict["network"]
+        self.add_marker(pytest.mark.fork(self.fork_name))
+        self.add_marker("evm_tools")
```
We should probably get rid of the `evm_tools` marker. This was used back when we needed to distinguish between our blockchain tests and state test runs. For context, we were using blockchain tests to test our specs directly and state tests to test the `t8n` implementation (hence the `evm_tools` name).

Also including the `evm_tools` marker here means that we mark all blockchain tests under json infra. This has the side effect of not running blockchain tests under the `optimized` environment. This explains why so few tests are run under `optimized`. See [this line](https://github.com/marioevz/execution-specs/blob/011051133789d4e4057503ce983e9cbbf11b1035/tox.ini#L100)

gurukamath replied:
Updated

gurukamath commented on tests/json_infra/helpers/load_blockchain_tests.py:
```diff
@@ -133,96 +58,150 @@ def add_block_to_chain(
+        )
+
+        expected_post_state = load.json_to_state(json_data["postState"])
+        assert chain.state == expected_post_state
```
I think this is currently not set up to catch any tests where the blocks themselves do not throw any exceptions but the overall state comparison fails . This I think is causing the current CI failure

gurukamath reviewed (commented):
Even though this is a much larger re-factor than #1730, I do like this approach since it uses more of the pytest native patterns. So a one-time larger change might be worth it.

gurukamath committed:
Run selective tests based on changed files (#1)

* fix(tests): remove evm_tools marker from blockchain tests

* remove coverage from json_infra

* enhance(tools): add json_test_name to Hardfork

* fix(tests): handle failing transactions in state tests

* enhance(tests): add from and until fork option to json_infra

* enhance(tests): run json_infra selectively

* enhance(tests): subclass Hardfork

* bug(tests): run all tests for t8n changes

* enhance(tests): minor fix

gurukamath committed:
fix(tests): ignore expectSection tests and add coverage

gurukamath committed:
enhance(tests): refactor exception markers

This commit refactors exception markers and marks the EEST static tests as slow

gurukamath commented:
This PR is almost ready for review. However, I'm moving this to draft in order to resolve the discrepancy between the collectd vs run tests in CI for `json_infra`.

gurukamath committed:
fix(tests): provide unique name to tests

gurukamath commented:
@SamWilsn This is now ready for review. The failing tests `json_infra` tests are unrelated and should be fixed with #1813

SamWilsn commented on .github/workflows/test.yaml:
```diff
@@ -105,13 +105,35 @@ jobs:
+
+          # Get changed files and save to disk
+          FILE_LIST="changed_files.txt"
+          git diff --name-only "$BASE_SHA" "$HEAD_SHA" > "$FILE_LIST"
```
Is `BASH_SHA` going to be the head of the base branch, or the merge-base of the two branches?

gurukamath replied:
`BASE_SHA` in this case would be the head of the base branch

SamWilsn replied:
Hm, so a file added in the base branch would get tested here, even if no changes were made to it in this pull request?

gurukamath replied:
The added file itself would not be tested but the addition might trigger a broader set of tests than what the PR explicitly changes. Perhaps this is not desirable and we should stick to comparing with the merge-base of the two branches. I'll give it a bit more thought

SamWilsn commented on .github/workflows/test.yaml:
```diff
@@ -105,13 +105,35 @@ jobs:
+          cat $FILE_LIST
       - name: Run json infra tests
-        run: tox -e json_infra
+        run: tox -e json_infra -- --file-list=${{ steps.get-changed-files.outputs.file_list }}
```
```suggestion
        run: tox -e json_infra -- --file-list="${{ steps.get-changed-files.outputs.file_list }}"
```

maybe?

gurukamath replied:
Updated

SamWilsn commented on tests/json_infra/conftest.py:
```diff
@@ -66,10 +74,33 @@ def pytest_addoption(parser: Parser) -> None:
-        dest="fork",
+        "--from",
+        action="store",
+        dest="forks_from",
```
Would it make sense to make a range class and store `fork`/`until`/`from` in the same `dest`?

gurukamath replied:
I'm ok with either approach on this. The current method is seems simple and straightforward to me. Whereas a single destination might make it easier to handle mutual exclusivity of the options. No strong leaning either way for me.

SamWilsn replied:
We can keep it the way it is.

SamWilsn commented on tests/json_infra/conftest.py:
```diff
@@ -66,10 +74,33 @@ def pytest_addoption(parser: Parser) -> None:
+        "--file-list",
+        action="store",
+        dest="file_list",
+        help="Only fill tests for the specified fork.",
```
Needs an update?

gurukamath replied:
Ah. Yes. Updated

SamWilsn commented on tests/json_infra/helpers/fixtures.py:
```diff
@@ -0,0 +1,118 @@
+
+    def clear_data_cache(self) -> None:
+        """Drop the data cache."""
+        self.__dict__.pop("data", None)
```
Would [this](https://stackoverflow.com/a/62662941) work/be more idiomatic?

gurukamath replied:
Updated

SamWilsn commented on tests/json_infra/helpers/select_tests.py:
```diff
@@ -0,0 +1,88 @@
+
+    # Extract affected forks
+    affected_forks = set()
+    repo_root = Path.cwd()  # Assuming running from repo root
```
You can get the root from pytest.

gurukamath replied:
Updated

SamWilsn reviewed (commented):
Partial review so far

gurukamath committed:
fix(tests): post review changes

gurukamath committed:
fix(tests): set BASE_SHA to merge base
